partner_multi_name/views/res_partner_views.xml

```diff
@@ -0,0 +1,69 @@
+
+            <xpath expr="//field[@name='lastname']" position="attributes">
+                <attribute name="attrs">{'required': ['|', '|', ('firstname', '!=', False), ('othernames', '!=', False), ('firstname', '=', False), ('othernames', '=', False), ('lastname2', '=', False), ('is_company', '=', False)]}</attribute>
+            </xpath>
+
+            <xpath expr="//field[@name='lastname2']" position="attributes">
+                <attribute name="attrs">{'required': ['|', '|', ('firstname', '!=', False), ('othernames', '!=', False), ('firstname', '=', False), ('othernames', '=', False), ('lastname', '=', False), ('is_company', '=', False)]}</attribute>
+            </xpath>
+        </field>
+    </record>
+
+    <record id="partner_form" model="ir.ui.view">
+        <field name="model">res.partner</field>
+        <field name="inherit_id" ref="partner_second_lastname.partner_form"/>
+        <field name="priority">100</field>
+        <field name="arch" type="xml">
+            <!-- Main form -->
+            <xpath expr="//field[@name='firstname']" position="replace"/>
+
+            <xpath expr="//field[@name='lastname']" position="before">
+                 <field name="firstname"
+                    attrs="{'required': [('othernames', '=', False), ('is_company', '=', False)]}"/>
+                <field name="othernames"
+                    attrs="{'required': [('firstname', '=', False), ('is_company', '=', False)]}"/>
+            </xpath>
+
+            <xpath expr="//field[@name='lastname']" position="attributes">
+                <attribute name="attrs">{'required': [('lastname2', '=', False), ('is_company', '=', False)]}</attribute>
+            </xpath>
+
+            <xpath expr="//field[@name='lastname2']" position="attributes">
+                <attribute name="attrs">{'required': [('lastname', '=', False), ('is_company', '=', False)]}</attribute>
+            </xpath>
+                
+            <!-- Inner contact form of child_ids -->
+            <xpath expr="//field[@name='child_ids']/form//field[@name='firstname']" position="replace"/>
+
+            <xpath expr="//field[@name='child_ids']/form//field[@name='lastname']" position="before">
+                <field name="firstname"
+                    attrs="{'required': [('othernames', '=', False), ('is_company', '=', False)]}"/>
+                <field name="othernames"
+                    attrs="{'required': [('firstname', '=', False), ('is_company', '=', False)]}"/>
+            </xpath>
+
+            <xpath expr="//field[@name='child_ids']/form//field[@name='lastname']" position="attributes">
+                <attribute name="attrs">{'required': [('lastname2', '=', False), ('is_company', '=', False)]}</attribute>
+            </xpath>
+
+            <xpath expr="//field[@name='child_ids']/form//field[@name='lastname2']" position="attributes">
+                <attribute name="attrs">{'required': [('lastname', '=', False), ('is_company', '=', False)]}</attribute>
```

Comment:
I don't think these are right. Compare to https://github.com/OCA/partner-contact/blob/10.0/partner_second_lastname/views/res_partner.xml#L31, then for the `firstname` field I think you'd want `{'required': [('othernames', '=', False), ('lastname', '=', False), ('lastname2', '=', False), ('is_company', '=', False)]}`. And so on for the others. So for each field you check the other three fields.